Add temporary pivot for rotating multiple 2D nodes

KoBeWi opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/3684
![godot windows tools 64_zJr9fwPUZV](https://user-images.githubusercontent.com/2223172/154871426-c22f6b17-05c8-4294-a8a2-f5cbf798bf2b.gif)


You can set a temporary pivot, which will be used as rotation origin for the selected nodes, in global coordinates. It's very useful when e.g. placing objects on your level (using a parent node as rotation point is very inconvenient, because usually you don't need it after rotation). The pivot disappears when selection changes.

~~Submitting as draft, as I didn't know where to put it xd Right now it uses hard-coded P shortcut. I've been thinking about 3 options:~~ Resolved.
- add a "Rotation Pivot" tool that would set the temporary pivot
- use the existing Pivot Tool and just make it so if you have multiple nodes selected, the temporary pivot would be set instead (which means you wouldn't be able to change pivot of multiple nodes)
- add a modifier for Pivot Tool, e.g. Alt. Then Alt+V would set the temporary pivot in Select mode

I like the third one the best, but it's probably up to discussion. The code uses the easiest option until this is settled.

KoBeWi committed:
Add temporary pivot for rotating multiple 2D nodes

KoBeWi commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -1277,6 +1277,14 @@ bool CanvasItemEditor::_gui_input_pivot(const Ref<InputEvent> &p_event) {
 		}
 	}
 
+	if (drag_type == DRAG_NONE) {
+		if (k.is_valid() && k->is_pressed() && !k->is_echo() && k->get_keycode() == Key::P && tool == TOOL_SELECT && k->get_modifiers_mask() == Key::NONE) {
+			temp_pivot = transform.affine_inverse().xform(viewport->get_local_mouse_position());
+			viewport->update();
+			return true;
+		}
+	}
```
This is the hard-coded part, so ignore it for now.

KoBeWi commented:
Ok I went with the third option. Alt with Pivot Tool moves temporary pivot (I had to make the tool active for any CanvasItem) and Alt + V changes temp pivot for selection tool.

akien-mga commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -1398,7 +1419,9 @@ bool CanvasItemEditor::_gui_input_rotate(const Ref<InputEvent> &p_event) {
 					drag_from = transform.affine_inverse().xform(b->get_position());
 					CanvasItem *canvas_item = drag_selection[0];
-					if (canvas_item->_edit_use_pivot()) {
+					if (temp_pivot != Vector2(INFINITY, INFINITY)) {
```
I'm not sure you can rely on this to work in a floating point comparison. That's why we have to use things like `isinf()` instead of equality operators for scalars, so that's probably the same for vectors.

KoBeWi commented:
> When you have multiple nodes selected, there is no visual feedback for the pivot location in the 2D editor after setting its position, nor whether it's permanent or temporary.

Fixed, I think. I couldn't really reproduce the problem, but I noticed the pivot did not appear when you click, only when you drag.

> The Alt modifier conflicts with window manipulation keys on many Linux window managers. I suggest adding an alternative.

Not this again 🙄 I think we decided that we don't care about window managers doing stupid things with Alt.
What alternative would you suggest? Note that the temp pivot works with selection tool, i.e. you can use Alt+V in that mode. Which means that either the pivot tool needs a new shortcut (which would make it inconsistent with the select shortcuts) or both tools need an alternative. Ctrl is not possible (because Ctrl+V) and Shift... idk.

> When selecting multiple nodes, could there be a quick way to place the pivot in the center of all selected nodes?

I added an Alt action when you press the Pivot tool. Kind of weird, but it works.

> Likewise, in the rotate tool, there could be a modifier to ignore the rotation pivots and rotate all selected nodes individually.

I don't think it's worth it, as it complicates things. Shortcut before dragging (e.g. hold Ctrl before starting rotating) would make it inconsistent with select tool and shortcut while rotating (e.g. hold Shift during rotation action) would result in toggling the pivot mid-rotation, which would either teleport the nodes or remember their position, which again is a hassle to handle xd
Removing the pivot involves changing node selection, so I think it's not really a problem.

KoBeWi commented:
Changed the modifier to Shift (it doesn't conflict with anything, so whatevs) and fixed a crash.

akien-mga commented:
Tested and this seems to work well.

The discoverability isn't the best, but that might come from me not actually knowing much of the existing hotkeys and tools.

I'm not sure how to use the current Pivot tool xD
But I did manage to place the temp pivot after selecting it and clicking with Shift, though it's a bit of a bother to then have to select the rotation tool manually to use it.

And from reading the PR I saw that Shift+V when in Select mode does place the temp pivot, which is nice. But is just "V" supposed to place the main pivot? Because that doesn't seem to work for me. I guess it may not work in multi-select?

The Shift+V behavior could be added to the (already lengthy) tooltip for Select Mode.

akien-mga reviewed (approved):
Works well for the most part, might need some iteration on the UX / discoverability.

KoBeWi commented:
> But is just "V" supposed to place the main pivot?

Yes:

https://github.com/godotengine/godot/assets/2223172/865adc05-5b6c-43f4-8408-99daa0cb0df9

Changing pivot involves changing node's position and setting the pivot so that the visual part stays in place. The pivot is not displayed when multiple nodes are selected, but setting it still works.

> The Shift+V behavior could be added to the (already lengthy) tooltip for Select Mode.

Shift+V is "inherited" from pivot tool. It's the same as Shift in scale tool, which is only mentioned in the scale tooltip, but you can use it in select mode too.

akien-mga commented:
Ok changing the pivot with "V" seems to work with a Sprite2D, but doesn't work with a CharacterBody2D. How come?

akien-mga commented:
Right, that makes sense. But that's still confusing to me, as if I multi-select 3 CharacterBody2D's, I can give them a temp pivot and rotate them fine :D
Likewise with a single selection, the temp pivot can be set, but the normal pivot can't.

akien-mga commented:
Thanks!
